samplers/aws/xray/internal/manifest.go

```diff
@@ -0,0 +1,351 @@
+			return false, fmt.Errorf("sampling statistics returned 5xx")
+		}
+
+		// set refresh flag if any sampling statistics returned 4xx
+		if strings.HasPrefix(*s.ErrorCode, "4") {
+			refresh = true
+		}
+	}
+
+	// set refresh flag if modifiedAt timestamp from remote is greater than ours
+	if remote := targets.LastRuleModification; remote != nil {
+		// convert unix timestamp to time.Time
+		lastRuleModification := time.Unix(int64(*targets.LastRuleModification), 0)
+
+		if lastRuleModification.After(m.refreshedAt) {
+			refresh = true
+		}
+	}
+
+	return
+}
+
+func (m *Manifest) updateReservoir(t *samplingTargetDocument) (err error) {
+	if t.RuleName == nil {
+		return fmt.Errorf("invalid sampling target. Missing rule name")
+	}
+
+	if t.FixedRate == nil {
+		return fmt.Errorf("invalid sampling target for rule %s. Missing fixed rate", *t.RuleName)
+	}
+
+	for index := range m.Rules {
+		if m.Rules[index].ruleProperties.RuleName == *t.RuleName {
+			m.Rules[index].reservoir.refreshedAt = m.Clock.Now()
+
+			// Update non-optional attributes from response
+			m.Rules[index].ruleProperties.FixedRate = *t.FixedRate
+
+			// Update optional attributes from response
+			if t.ReservoirQuota != nil {
+				m.Rules[index].reservoir.quota = *t.ReservoirQuota
+			}
+			if t.ReservoirQuotaTTL != nil {
+				m.Rules[index].reservoir.expiresAt = time.Unix(int64(*t.ReservoirQuotaTTL), 0)
+			}
+			if t.Interval != nil {
+				m.Rules[index].reservoir.interval = time.Duration(*t.Interval)
+			}
+		}
+	}
+
+	return
+}
+
+// snapshots takes a snapshot of sampling statistics from all rules, resetting
+// statistics counters in the process.
+func (m *Manifest) snapshots() ([]*samplingStatisticsDocument, error) {
+	statistics := make([]*samplingStatisticsDocument, 0, len(m.Rules)+1)
+
+	// Generate sampling statistics for user-defined rules
+	for index := range m.Rules {
+		if m.Rules[index].stale(m.Clock.Now()) {
+			s := m.Rules[index].snapshot(m.Clock.Now())
+			s.ClientID = m.clientID
+
+			statistics = append(statistics, s)
+		}
+	}
+
+	return statistics, nil
+}
+
+// sort sorts the rule array first by priority and then by rule name.
+func (m *Manifest) sort() {
+	// comparison function
+	less := func(i, j int) bool {
+		if m.Rules[i].ruleProperties.Priority == m.Rules[j].ruleProperties.Priority {
+			return strings.Compare(m.Rules[i].ruleProperties.RuleName, m.Rules[j].ruleProperties.RuleName) < 0
+		}
+		return m.Rules[i].ruleProperties.Priority < m.Rules[j].ruleProperties.Priority
+	}
+
+	sort.Slice(m.Rules, less)
+}
+
+// minimumPollingInterval finds the minimum interval amongst all the targets
+func (m *Manifest) minimumPollingInterval() (minPoll time.Duration) {
+	minPoll = time.Duration(0)
+	for _, rules := range m.Rules {
+		if minPoll == 0 {
+			minPoll = rules.reservoir.interval
+		} else {
+			if minPoll >= rules.reservoir.interval {
+				minPoll = rules.reservoir.interval
+			}
+		}
+	}
+
+	return minPoll * time.Second
+}
```

Comment:
```suggestion
func (m *Manifest) minimumPollingInterval() time.Duration {
	minPoll = math.MaxInt
	for _, rules := range m.Rules {
		if minPoll > rules.reservoir.interval {
			minPoll = rules.reservoir.interval
		}
	}

	return minPoll * time.Second
}
```

This could still end up at zero, so not sure having two separate branches does much good.  If you want to return `0` in the case of an empty list then do that as a check up-front:

```go
if len(m.Rules) == 0 {
  return time.Duration(0)
}
```

Reply:
Okay I will update the logic like below. if we initialize `minPoll = math.MaxInt` then caller has to check 2 conditions `minPoll != math.MaxInt` and `minPoll != 0` because we don't want to update `m.SamplingTargetsPollingInterval` with zero or `math.MaxInt` in that case we would rather not update. So if we initialize with 0 then we can achieve this by only checking `minPoll != 0` and we can update the `m.SamplingTargetsPollingInterval`

```
	minPoll = time.Duration(0)
	for _, rules := range m.Rules {
		if minPoll > rules.reservoir.interval {
			minPoll = rules.reservoir.interval
		}
	}
	return minPoll * time.Second
```

Reply:
```go
	minPoll = time.Duration(0)
	for _, rules := range m.Rules {
		if minPoll > rules.reservoir.interval {
			minPoll = rules.reservoir.interval
		}
	}
	return minPoll * time.Second
```

This might as well be:

```go
return time.Duration(0)
```

If, instead, you use this:

```go
func (m *Manifest) minimumPollingInterval() time.Duration {
	if len(m.Rules) == 0 {
		return time.Duration(0)
	}

	minPoll = math.MaxInt
	for _, rules := range m.Rules {
		if minPoll > rules.reservoir.interval {
			minPoll = rules.reservoir.interval
		}
	}
	return minPoll * time.Second
}
```

It will only return `time.Duration(0)` if there are no rules and will never return `math.MaxInt` as there will be at least one rule and that rule's interval will be less than `math.MaxInt` in any sane universe.  If it isn't, then that truly is the minimum interval and so be it.